code/datums/components/material_container.dm

```diff
@@ -116,82 +116,161 @@
+		if(original_item != active_held)
+			original_item = null
+		if(!isnull(original_item) && !user.temporarilyRemoveItemFromInventory(original_item)) //remove from hand(if split remove the original stack else the target)
+			to_chat(user, span_warning("[held_item] is stuck to you and cannot be placed into [parent]."))
 			return
-	if(!user.temporarilyRemoveItemFromInventory(held_item))
-		to_chat(user, span_warning("[held_item] is stuck to you and cannot be placed into [parent]."))
-		return
-	var/inserted = insert_item(held_item, stack_amt = requested_amount, breakdown_flags= mat_container_flags)
-	if(inserted)
-		to_chat(user, span_notice("You insert a material total of [inserted] into [parent]."))
-		qdel(held_item)
-		if(after_insert)
-			after_insert.Invoke(held_item, last_inserted_id, inserted)
-	else if(held_item == active_held)
-		user.put_in_active_hand(held_item)
 
-/// Proc specifically for inserting items, returns the amount of materials entered.
-/datum/component/material_container/proc/insert_item(obj/item/I, multiplier = 1, stack_amt, breakdown_flags = mat_container_flags)
-	if(QDELETED(I))
-		return FALSE
+		//insert the item
+		var/inserted = insert_item(target, breakdown_flags = mat_container_flags)
+		if(inserted > 0)
+			. += inserted
+
+			//stack was either split by the container(!QDELETED(target) means the container only consumed a part of it) or by the player, put whats left back of the original stack back in players hand
+			if((!QDELETED(target) || was_stack_split))
+
+				//stack was split by player and that portion was not fully consumed, merge whats left back with the original stack
+				if(!QDELETED(target) && was_stack_split)
+					var/obj/item/stack/inserting_stack = target
+					item_stack.add(inserting_stack.amount)
+					qdel(inserting_stack)
+
+				//was this the original item in the players hand? put what's left back in the player's hand
+				if(!isnull(original_item))
+					user.put_in_active_hand(original_item)
 
+			to_chat(user, span_notice("You insert a material total of [inserted] into [parent]."))
+		else
+			//decode the error & print it
+			var/error_msg
+			if(inserted == -2)
+				error_msg = "[parent] has insufficient space to accept the [target]"
+			else
+				error_msg = "[target] has insufficient materials to be accepted by [parent]"
+			to_chat(user, span_warning(error_msg))
+
+			//player split the stack by the requested amount but even that split amount could not be salvaged. merge it back with the original
+			if(!isnull(item_stack) && was_stack_split)
+				var/obj/item/stack/inserting_stack = target
+				item_stack.add(inserting_stack.amount)
+				qdel(inserting_stack)
+
+			//was this the original item in the players hand? put it back because we coudn't salvage it
+			if(!isnull(original_item))
+				user.put_in_active_hand(original_item)
+
+//splits a stack. we don't use /obj/item/stack/proc/split_stack because Byond complains that should only be called asynchronously
```

Comment:
You can invoke it async, can't you?

If this is just a copy paste proc then I think that would be the better solution.

Reply:
We can't invoke it asynchronously because we need the immediate return value of this proc if you see the code.

Also its not a copy paste it's a very slimed down version of it if you compare it to the original stack split proc 
https://github.com/tgstation/tgstation/blob/9ecce38f237c9b92b4727c93f3185899bf37459d/code/game/objects/items/stacks/stack.dm#L674-L687

The differences are
1. Removed user/mob interaction interaction(which is what byond complained because that was asynchronous, so even passing null the compiler would still complain cause it doesn't know that)
2. Removed adding finger prints to the newly split stack cause it will be deleted anyway 
3. No need to referesh the atom local storage.

So comparatively this proc is much faster